Intermittent failure of CamelDevModeTest

Seems `CamelMainEventBridge` fires events after Quarkus is shutdown and thus the Arc container is `null`. Hence `IllegalArgumentException` is thrown [here](https://github.com/quarkusio/quarkus/blob/61d80b88280345f4f653bc730abd6f814e84cd11/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java#L696).

```
[INFO] Running org.apache.camel.quarkus.main.CamelDevModeTest
2020-07-09 15:38:44,916 INFO  [org.apa.cam.qua.mai.dep.CamelMainHotDeploymentProcessor] (build-15) HotDeployment files (camel.main.xml-routes):
2020-07-09 15:38:44,916 INFO  [org.apa.cam.qua.mai.dep.CamelMainHotDeploymentProcessor] (build-15) - /tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:46,104 INFO  [org.apa.cam.qua.cor.CamelBootstrapRecorder] (Quarkus Main Thread) bootstrap runtime: org.apache.camel.quarkus.main.CamelMainRuntime
2020-07-09 15:38:46,136 INFO  [org.apa.cam.mai.BaseMainSupport] (Quarkus Main Thread) Auto-configuration summary:
2020-07-09 15:38:46,137 INFO  [org.apa.cam.mai.BaseMainSupport] (Quarkus Main Thread) 	camel.main.xmlRoutes=file:/tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:46,139 INFO  [org.apa.cam.qua.mai.CamelMainRoutesCollector] (Quarkus Main Thread) Loading additional Camel XML routes from: file:/tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:46,247 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-1) is starting
2020-07-09 15:38:46,249 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) StreamCaching is not in use. If using streams then its recommended to enable stream caching. See more details at http://camel.apache.org/stream-caching.html
2020-07-09 15:38:46,253 INFO  [org.apa.cam.imp.eng.InternalRouteStartupManager] (Quarkus Main Thread) Route: r1 started and consuming from: direct://start
2020-07-09 15:38:46,255 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Total 1 routes, of which 1 are started
2020-07-09 15:38:46,255 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-1) started in 0.007 seconds
2020-07-09 15:38:46,341 INFO  [io.quarkus] (Quarkus Main Thread) Quarkus 1.6.0.Final on JVM started in 2.007s. Listening on: http://0.0.0.0:8080
2020-07-09 15:38:46,341 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-07-09 15:38:46,341 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [camel-core, camel-direct, camel-log, camel-main, camel-policy, camel-support-common, camel-timer, camel-xml-io, cdi, resteasy, resteasy-jsonb]
2020-07-09 15:38:48,400 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (Timer-0) File change detected: /tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:48,403 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-1) is shutting down
2020-07-09 15:38:48,408 INFO  [org.apa.cam.mai.MainLifecycleStrategy] (Quarkus Main Thread) CamelContext: camel-1 has been shutdown, triggering shutdown of the JVM.
2020-07-09 15:38:48,410 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-1) uptime 2.161 seconds
2020-07-09 15:38:48,410 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-1) is shutdown in 0.007 seconds
2020-07-09 15:38:48,442 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-2) File change detected: /tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:48,457 INFO  [io.quarkus] (Quarkus Main Thread) Quarkus stopped in 0.055s
2020-07-09 15:38:48,555 INFO  [org.apa.cam.qua.mai.dep.CamelMainHotDeploymentProcessor] (build-24) HotDeployment files (camel.main.xml-routes):
2020-07-09 15:38:48,555 INFO  [org.apa.cam.qua.mai.dep.CamelMainHotDeploymentProcessor] (build-24) - /tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:48,828 INFO  [org.apa.cam.qua.cor.CamelBootstrapRecorder] (Quarkus Main Thread) bootstrap runtime: org.apache.camel.quarkus.main.CamelMainRuntime
2020-07-09 15:38:48,837 INFO  [org.apa.cam.mai.BaseMainSupport] (Quarkus Main Thread) Auto-configuration summary:
2020-07-09 15:38:48,837 INFO  [org.apa.cam.mai.BaseMainSupport] (Quarkus Main Thread) 	camel.main.xmlRoutes=file:/tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:48,838 INFO  [org.apa.cam.qua.mai.CamelMainRoutesCollector] (Quarkus Main Thread) Loading additional Camel XML routes from: file:/tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:48,840 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-2) is starting
2020-07-09 15:38:48,841 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) StreamCaching is not in use. If using streams then its recommended to enable stream caching. See more details at http://camel.apache.org/stream-caching.html
2020-07-09 15:38:48,841 INFO  [org.apa.cam.imp.eng.InternalRouteStartupManager] (Quarkus Main Thread) Route: r2 started and consuming from: direct://start
2020-07-09 15:38:48,842 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Total 1 routes, of which 1 are started
2020-07-09 15:38:48,842 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-2) started in 0.002 seconds
2020-07-09 15:38:48,856 INFO  [io.quarkus] (Quarkus Main Thread) Quarkus 1.6.0.Final on JVM started in 0.391s. Listening on: http://0.0.0.0:8080
2020-07-09 15:38:48,856 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-07-09 15:38:48,857 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [camel-core, camel-direct, camel-log, camel-main, camel-policy, camel-support-common, camel-timer, camel-xml-io, cdi, resteasy, resteasy-jsonb]
2020-07-09 15:38:48,857 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (Timer-0) Hot replace total time: 0.457s 
2020-07-09 15:38:48,857 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-2) is shutting down
2020-07-09 15:38:48,859 INFO  [org.apa.cam.mai.MainLifecycleStrategy] (Quarkus Main Thread) CamelContext: camel-2 has been shutdown, triggering shutdown of the JVM.
2020-07-09 15:38:48,859 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-2) uptime 0.019 seconds
2020-07-09 15:38:48,860 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-2) is shutdown in 0.003 seconds
2020-07-09 15:38:48,860 INFO  [io.quarkus] (Quarkus Main Thread) Quarkus stopped in 0.003s
2020-07-09 15:38:48,861 ERROR [org.apa.cam.qua.mai.CamelMainRuntime] (Thread-265) Failed to start application: java.lang.IllegalArgumentException
	at io.quarkus.arc.impl.ArcContainerImpl.unwrap(ArcContainerImpl.java:697)
	at io.quarkus.arc.impl.EventImpl.createNotifier(EventImpl.java:138)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at io.quarkus.arc.impl.EventImpl.getNotifier(EventImpl.java:112)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:69)
	at org.apache.camel.quarkus.main.CamelMainEventBridge.fireEvent(CamelMainEventBridge.java:89)
	at org.apache.camel.quarkus.main.CamelMainEventBridge.afterStop(CamelMainEventBridge.java:85)
	at org.apache.camel.main.MainSupport.afterStop(MainSupport.java:118)
	at org.apache.camel.quarkus.main.CamelMain.runEngine(CamelMain.java:138)
	at org.apache.camel.quarkus.main.CamelMainRuntime.lambda$start$0(CamelMainRuntime.java:50)
	at java.base/java.lang.Thread.run(Thread.java:834)

2020-07-09 15:38:48,944 INFO  [org.apa.cam.qua.mai.dep.CamelMainHotDeploymentProcessor] (build-12) HotDeployment files (camel.main.xml-routes):
2020-07-09 15:38:48,944 INFO  [org.apa.cam.qua.mai.dep.CamelMainHotDeploymentProcessor] (build-12) - /tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:49,207 INFO  [org.apa.cam.qua.cor.CamelBootstrapRecorder] (Quarkus Main Thread) bootstrap runtime: org.apache.camel.quarkus.main.CamelMainRuntime
2020-07-09 15:38:49,216 INFO  [org.apa.cam.mai.BaseMainSupport] (Quarkus Main Thread) Auto-configuration summary:
2020-07-09 15:38:49,216 INFO  [org.apa.cam.mai.BaseMainSupport] (Quarkus Main Thread) 	camel.main.xmlRoutes=file:/tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:49,216 INFO  [org.apa.cam.qua.mai.CamelMainRoutesCollector] (Quarkus Main Thread) Loading additional Camel XML routes from: file:/tmp/camel-devmode-9461026744410861234/routes.xml
2020-07-09 15:38:49,219 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-3) is starting
2020-07-09 15:38:49,220 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) StreamCaching is not in use. If using streams then its recommended to enable stream caching. See more details at http://camel.apache.org/stream-caching.html
2020-07-09 15:38:49,221 INFO  [org.apa.cam.imp.eng.InternalRouteStartupManager] (Quarkus Main Thread) Route: r2 started and consuming from: direct://start
2020-07-09 15:38:49,222 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Total 1 routes, of which 1 are started
2020-07-09 15:38:49,222 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-3) started in 0.002 seconds
2020-07-09 15:38:49,224 INFO  [io.quarkus] (Quarkus Main Thread) Quarkus 1.6.0.Final on JVM started in 0.358s. Listening on: http://0.0.0.0:8080
2020-07-09 15:38:49,225 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-07-09 15:38:49,225 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [camel-core, camel-direct, camel-log, camel-main, camel-policy, camel-support-common, camel-timer, camel-xml-io, cdi, resteasy, resteasy-jsonb]
2020-07-09 15:38:49,226 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-2) Hot replace total time: 0.785s 
2020-07-09 15:38:49,252 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-3) is shutting down
2020-07-09 15:38:49,254 INFO  [org.apa.cam.mai.MainLifecycleStrategy] (Quarkus Main Thread) CamelContext: camel-3 has been shutdown, triggering shutdown of the JVM.
2020-07-09 15:38:49,255 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-3) uptime 0.034 seconds
2020-07-09 15:38:49,255 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (Quarkus Main Thread) Apache Camel 3.4.0 (camel-3) is shutdown in 0.003 seconds
2020-07-09 15:38:49,255 INFO  [io.quarkus] (Quarkus Main Thread) Quarkus stopped in 0.003s
Exception in thread "Thread-289" java.lang.RuntimeException: java.lang.IllegalArgumentException
	at org.apache.camel.quarkus.main.CamelMainRuntime.lambda$start$0(CamelMainRuntime.java:54)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalArgumentException
	at io.quarkus.arc.impl.ArcContainerImpl.unwrap(ArcContainerImpl.java:697)
	at io.quarkus.arc.impl.EventImpl.createNotifier(EventImpl.java:138)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at io.quarkus.arc.impl.EventImpl.getNotifier(EventImpl.java:112)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:69)
	at org.apache.camel.quarkus.main.CamelMainEventBridge.fireEvent(CamelMainEventBridge.java:89)
	at org.apache.camel.quarkus.main.CamelMainEventBridge.afterStop(CamelMainEventBridge.java:85)
	at org.apache.camel.main.MainSupport.afterStop(MainSupport.java:118)
	at org.apache.camel.quarkus.main.CamelMain.runEngine(CamelMain.java:138)
	at org.apache.camel.quarkus.main.CamelMainRuntime.lambda$start$0(CamelMainRuntime.java:50)
	... 1 more
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 6.903 s <<< FAILURE! - in org.apache.camel.quarkus.main.CamelDevModeTest
[ERROR] testRoutesDiscovery  Time elapsed: 6.867 s  <<< ERROR!
java.lang.RuntimeException: java.lang.IllegalArgumentException
Caused by: java.lang.IllegalArgumentException
```

--- a/extensions-core/main/deployment/src/main/java/org/apache/camel/quarkus/main/deployment/CamelMainProcessor.java
+++ b/extensions-core/main/deployment/src/main/java/org/apache/camel/quarkus/main/deployment/CamelMainProcessor.java
@@ -42,6 +42,7 @@ import org.apache.camel.quarkus.core.deployment.spi.CamelRuntimeTaskBuildItem;
 import org.apache.camel.quarkus.core.deployment.spi.ContainerBeansBuildItem;
 import org.apache.camel.quarkus.core.deployment.spi.RuntimeCamelContextCustomizerBuildItem;
 import org.apache.camel.quarkus.main.CamelMain;
+import org.apache.camel.quarkus.main.CamelMainConfig;
 import org.apache.camel.quarkus.main.CamelMainProducers;
 import org.apache.camel.quarkus.main.CamelMainRecorder;
 import org.apache.camel.quarkus.main.deployment.spi.CamelMainBuildItem;
@@ -140,14 +141,15 @@ public class CamelMainProcessor {
      * after having processed a certain number of messages..
      * </ul>
      *
-     * @param  index         a reference to a {@link IndexView}
-     * @param  beanContainer a reference to a fully initialized CDI bean container
-     * @param  recorder      the recorder.
-     * @param  main          a reference to a {@link CamelMain}.
-     * @param  customizers   a list of {@link org.apache.camel.quarkus.core.CamelContextCustomizer} that will be
-     *                       executed before starting the {@link CamelContext} at {@link ExecutionTime#RUNTIME_INIT}.
-     * @param  runtimeTasks  a placeholder to ensure all the runtime task are properly are done.
-     * @return               a build item holding a {@link CamelRuntime} instance.
+     * @param  index           a reference to a {@link IndexView}
+     * @param  beanContainer   a reference to a fully initialized CDI bean container
+     * @param  recorder        the recorder.
+     * @param  main            a reference to a {@link CamelMain}.
+     * @param  customizers     a list of {@link org.apache.camel.quarkus.core.CamelContextCustomizer} that will be
+     *                         executed before starting the {@link CamelContext} at {@link ExecutionTime#RUNTIME_INIT}.
+     * @param  runtimeTasks    a placeholder to ensure all the runtime task are properly are done.
+     * @param  camelMainConfig a {@link CamelMainConfig}
+     * @return                 a build item holding a {@link CamelRuntime} instance.
      */
     @BuildStep
     @Record(value = ExecutionTime.RUNTIME_INIT, optional = true)
@@ -161,7 +163,8 @@ public class CamelMainProcessor {
             CamelMainRecorder recorder,
             CamelMainBuildItem main,
             List<RuntimeCamelContextCustomizerBuildItem> customizers,
-            List<CamelRuntimeTaskBuildItem> runtimeTasks) {
+            List<CamelRuntimeTaskBuildItem> runtimeTasks,
+            CamelMainConfig camelMainConfig) {
 
         // Run the customizer before starting the context to give a last chance
         // to amend the Camel Context setup.
@@ -174,7 +177,10 @@ public class CamelMainProcessor {
                 customizers.stream().map(RuntimeCamelContextCustomizerBuildItem::get).collect(Collectors.toList()));
 
         return new CamelRuntimeBuildItem(
-                recorder.createRuntime(beanContainer.getValue(), main.getInstance()),
+                recorder.createRuntime(
+                        beanContainer.getValue(),
+                        main.getInstance(),
+                        camelMainConfig.shutdown.timeout.toMillis()),
                 index.getIndex().getAnnotations(DotName.createSimple(QuarkusMain.class.getName())).isEmpty());
     }
 }

--- a/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMain.java
+++ b/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMain.java
@@ -34,8 +34,12 @@ import org.apache.camel.main.MainShutdownStrategy;
 import org.apache.camel.spi.CamelBeanPostProcessor;
 import org.apache.camel.spi.HasCamelContext;
 import org.apache.camel.support.service.ServiceHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public final class CamelMain extends MainCommandLineSupport implements HasCamelContext {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CamelMain.class);
+
     private final AtomicBoolean engineStarted;
 
     public CamelMain(CamelContext camelContext) {
@@ -182,7 +186,11 @@ public final class CamelMain extends MainCommandLineSupport implements HasCamelC
 
         @Override
         public void await(long timeout, TimeUnit unit) throws InterruptedException {
-            latch.await(timeout, unit);
+            if (!latch.await(timeout, unit)) {
+                LOGGER.warn(
+                        "Could not await stopping CamelMain within {} {}. You may want to increase camel.main.shutdown.timeout",
+                        timeout, unit);
+            }
         }
     }
 }

--- a/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMainConfig.java
+++ b/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMainConfig.java
@@ -16,5 +16,28 @@
  */
 package org.apache.camel.quarkus.main;
 
+import java.time.Duration;
+
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "camel.main", phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
 public class CamelMainConfig {
+
+    /**
+     * Build time configuration options for {@link CamelMain} shutdown.
+     */
+    @ConfigItem
+    public ShutdownConfig shutdown;
+
+    @ConfigGroup
+    public static class ShutdownConfig {
+        /**
+         * A timeout (with millisecond precision) to wait for {@link CamelMain#stop()} to finish
+         */
+        @ConfigItem(defaultValue = "PT3S")
+        public Duration timeout;
+    }
 }

--- a/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMainRecorder.java
+++ b/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMainRecorder.java
@@ -98,8 +98,11 @@ public class CamelMainRecorder {
         });
     }
 
-    public RuntimeValue<CamelRuntime> createRuntime(BeanContainer beanContainer, RuntimeValue<CamelMain> main) {
-        final CamelRuntime runtime = new CamelMainRuntime(main.getValue());
+    public RuntimeValue<CamelRuntime> createRuntime(
+            BeanContainer beanContainer,
+            RuntimeValue<CamelMain> main,
+            long shutdownTimeoutMs) {
+        final CamelRuntime runtime = new CamelMainRuntime(main.getValue(), shutdownTimeoutMs);
 
         // register to the container
         beanContainer.instance(CamelProducers.class).setRuntime(runtime);

--- a/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMainRuntime.java
+++ b/extensions-core/main/runtime/src/main/java/org/apache/camel/quarkus/main/CamelMainRuntime.java
@@ -17,6 +17,7 @@
 package org.apache.camel.quarkus.main;
 
 import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
 
 import io.quarkus.runtime.Quarkus;
 import org.apache.camel.CamelContext;
@@ -30,9 +31,11 @@ import org.slf4j.LoggerFactory;
 public class CamelMainRuntime implements CamelRuntime {
     private static final Logger LOGGER = LoggerFactory.getLogger(CamelMainRuntime.class);
     private final CamelMain main;
+    private final long shutdownTimeoutMs;
 
-    public CamelMainRuntime(CamelMain main) {
+    public CamelMainRuntime(CamelMain main, long shutdownTimeoutMs) {
         this.main = main;
+        this.shutdownTimeoutMs = shutdownTimeoutMs;
     }
 
     @Override
@@ -53,7 +56,7 @@ public class CamelMainRuntime implements CamelRuntime {
                     stop();
                     throw new RuntimeException(e);
                 }
-            }).start();
+            }, "camel-main").start();
         } catch (Exception e) {
             LOGGER.error("Failed to start application", e);
             stop();
@@ -64,6 +67,12 @@ public class CamelMainRuntime implements CamelRuntime {
     @Override
     public void stop() {
         main.stop();
+        /* Wait till the Camel shutdown is finished in camel-main thread started in start(String[]) above */
+        try {
+            main.getShutdownStrategy().await(shutdownTimeoutMs, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
     }
 
     @Override

